extensions/AWS/S3/AWSS3Storage.cs

```diff
@@ -0,0 +1,261 @@
+        string index,
+        string documentId,
+        string fileName,
+        bool logErrIfNotFound = true,
+        CancellationToken cancellationToken = default)
+    {
+        var objectKey = $"{index}/{documentId}/{fileName}";
+
+        try
+        {
+            GetObjectRequest request = new() { BucketName = this._bucketName, Key = objectKey };
+            var response = await this._client.GetObjectAsync(request, cancellationToken).ConfigureAwait(false);
+
+            var memoryStream = new MemoryStream();
+            await response.ResponseStream.CopyToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+
+            return new StreamableFileContent(
+                fileName,
+                response.ContentLength,
+                response.Headers.ContentType,
+                response.LastModified,
+                () =>
+                {
+                    memoryStream.Seek(0, SeekOrigin.Begin);
+                    return Task.FromResult((Stream)memoryStream);
+                }
+            );
+        }
+        catch (AmazonS3Exception e) when (e.StatusCode == HttpStatusCode.NotFound)
+        {
+            if (logErrIfNotFound)
+            {
+                this._log.LogInformation("File not found: {0}", objectKey);
+            }
+
+            throw new DocumentStorageFileNotFoundException("File not found", e);
+        }
+    }
+
+    /// <inheritdoc />
+    public void Dispose()
+    {
+        this._client.Dispose();
+    }
+
+    #region private
+
+    private async Task DeleteObjectsByPrefixAsync(string prefix, CancellationToken cancellationToken)
+    {
+        if (string.IsNullOrWhiteSpace(prefix))
+        {
+            throw new DocumentStorageException("The object prefix is empty, stopping the process to prevent data loss");
+        }
+
+        this._log.LogTrace("Deleting objects with prefix '{0}'", prefix);
+
+        var allObjects = new List<S3Object>();
+        var request = new ListObjectsV2Request
+        {
+            BucketName = this._bucketName,
+            Prefix = prefix
+        };
+
+        do
+        {
+            ListObjectsV2Response? response = await this._client.ListObjectsV2Async(
+                request,
+                cancellationToken: cancellationToken
+            ).ConfigureAwait(false);
+
+            if (response == null) { break; }
+
+            allObjects.AddRange(response.S3Objects);
+
+            if (!response.IsTruncated)
+            {
+                // Exit the loop if there are no more objects to retrieve
+                break;
+            }
+
+            request.ContinuationToken = response.NextContinuationToken;
+        } while (true);
+
+        foreach (var obj in allObjects)
+        {
+            var fileName = obj.Key.Trim('/').Substring(prefix.Trim('/').Length).Trim('/');
+
+            // Don't delete the pipeline status file
+            if (fileName == Constants.PipelineStatusFilename) { continue; }
```

Comment:
Just to be sure when deleting document from the database we delete everything from the storage except this file `__pipeline_status.json`, right? I saw this in the Azure Blob implementation. Btw, is there a reason to keep this file?

Reply:
deleting the file causes an error in the pipeline orchestration, it would take some rewrite to sort out so for now we settled on leaving it in place. If leaving these files behind becomes a problem the current suggestion is to have a periodic purge process out of band, identifying and deleting these folders.